pkg/kubelet/kubelet.go

```diff
@@ -1889,7 +1889,10 @@ func (kl *Kubelet) syncLoopIteration(configCh <-chan kubetypes.PodUpdate, handle
 
 		if e.Type == pleg.ContainerDied {
 			if containerID, ok := e.Data.(string); ok {
-				kl.cleanUpContainersInPod(e.ID, containerID)
+				cid := kubecontainer.ParseContainerID(containerID)
+				kl.cleanUpContainersInPod(e.ID, cid.ID)
+				// Update pod's conditions(Ready and ContainersReady) immediately.
+				kl.statusManager.SetContainerReadiness(e.ID, cid, false)
```

Comment:
I don't think we can do this from the main kubelet thread.  Though unlikely, there are race conditions with the pod worker thread IIRC.  We may need to find another way to do this in the pod worker or syncPod function...

Reply:
Thanks for the suggestion. I find it is a little hard to add the logic to pod worker or syncPod function. There is a lock for `SetContainerReadiness`, so it is safe to call it in kubelet thread?

Reply:
That is just to prevent races between calls to SetPodStatus and SetContainerReadiness.  It won't fix races between the kubelet's housekeeping thread and the pod worker's thread.